src/traces/scattermapbox/plot.js

```diff
@@ -101,29 +101,179 @@ proto.update = function update(calcTrace) {
+    });
+};
+
+scatterclustermapboxproto.setSourceData = function(opts) {
+    this.subplot.map
+        .getSource(this.sourceId)
+        .setData(opts.circle.geojson);
+};
+
+scatterclustermapboxproto.addLayer = function(opts, below) {
+    var clusterId = this.layerIds.cluster;
+    var sourceId = this.sourceId;
+    var map = this.subplot.map;
+    this.subplot.addLayer({
+        id: clusterId,
+        type: 'circle',
+        source: sourceId,
+        layout: opts.cluster.layout,
+        filter: ['has', 'point_count'],
+        paint: opts.cluster.paint
+    }, below);
+    this.subplot.addLayer({
+        id: this.layerIds.clusterCount,
+        type: 'symbol',
+        source: sourceId,
+        filter: ['has', 'point_count'],
+        layout: opts.clusterCount.layout
+    }, below);
+    this.subplot.addLayer({
+        id: this.layerIds.clusterPoints,
+        type: 'circle',
+        source: sourceId,
+        layout: opts.circle.layout,
+        paint: opts.circle.paint,
+        filter: ['!', ['has', 'point_count']],
+    }, below);
+    this.subplot.map.on('click', clusterId, function(e) {
+        var features = map.queryRenderedFeatures(e.point, {
+            layers: [clusterId]
+        });
+        var clusterId_ = features[0].properties.cluster_id;
+        map.getSource(sourceId).getClusterExpansionZoom(
+            clusterId_,
+            function(err, zoom) {
+                if(err) return;
+                map.easeTo({
+                    center: features[0].geometry.coordinates,
+                    zoom: zoom
+                });
+            }
+            );
+    });
+};
+
+scatterclustermapboxproto.update = function update(calcTrace) {
+    var subplot = this.subplot;
+    var map = subplot.map;
     var optsAll = convert(subplot.gd, calcTrace);
-    var below = scatterMapbox.below = subplot.belowLookup['trace-' + trace.uid];
+    var below = subplot.belowLookup['trace-' + this.uid];
+    var i, k, opts;
+
+    if(below !== this.below) {
+        for(i = this.order.length - 1; i >= 0; i--) {
+            k = this.order[i];
+            map.removeLayer(this.layerIds[k]);
+        }
+        this.addLayer(optsAll, below);
+        this.below = below;
+    }
+
+    for(i = 0; i < this.order.length; i++) {
+        k = this.order[i];
+        opts = optsAll[k];
 
-    for(var i = 0; i < ORDER.length; i++) {
-        var k = ORDER[i];
-        var opts = optsAll[k];
-        scatterMapbox.addSource(k, opts);
-        scatterMapbox.addLayer(k, opts, below);
+        subplot.setOptions(this.layerIds[k], 'setLayoutProperty', opts.layout);
+
+        if(opts.layout.visibility === 'visible') {
+            this.setSourceData(k, opts);
+            subplot.setOptions(this.layerIds[k], 'setPaintProperty', opts.paint);
+        }
     }
 
+    // link ref for quick update during selections
+    calcTrace[0].trace._glTrace = this;
+};
+
+scatterclustermapboxproto.dispose = function dispose() {
```

Comment:
Now that there are two prototypes to take care of, one should check if reacting `scattermapbox` with/without `cluster` works as expected.

Reply:
So basically, we need to ensure that the correct layers in mapbox is rendered depending on whether cluster has been assigned a value. Is that correctly understood?

Reply:
Exactly.
Here is a [codepen](https://codepen.io/MojtabaSamimi/pen/WNwKxeQ?editors=0010) illustrating the problem.
We should see the cluster after `Plotly.react` not the initial figure.

Reply:
This is now fixed by 3f6ce61.
Thanks @elben10.